### ISSUE TITLE
Improve handling of shared images

### ISSUE THREAD

**Issue by bmpotter** [open-horizon/anax#175]
To more easily discuss this we'll use the concrete example of the location poc and gpstest poc sharing a single instance of the gps container. The current handling of shared images/containers is:

- each poc policy file specifies the gps container as shared in the deployment string, but they each seed the gps image separately into their own torrent file.
- when anax downloads the gps image tar file for the location workload, it uses `docker load` to create the image, and then starts the container on a shared network.
- when anax processes the gpstest policy file, it doesn't immediately recognize the gps torrent tar file as the same image, so it goes thru its normal processing: downloads the gps image tar file and uses `docker load` to create the image. Docker load does nothing if the image layers are all the same. At this point anax recognizes that this shared image is already running, so it doesn't have to start it.

The downside of the implementation above is that if the gps container is shared between all the pocs, anax could end up downloading the same gps image 4 or 5 times. To avoid this, there are (at least) 2 options:

### Option 1: Anax Recognizes That Two Tar Files Are the Same Image

- when anax downloads and docker loads an image, keep a mapping between the tar file name and docker image name
- when anax encounters the 2nd workload that references the gps image name as shared, find the tar file it already has for that image and check all of the image signatures in the torrent section of the policy file against that tar file. For the 1 that matches, anax will know it doesn't have to download that file. (Does anax download each tar file individually, or the whole bundle at once using the torrent url?? Both. It downloads the torrent file, via HTTP GET, and then downloads the images referenced within the torrent file.)

### Option 2: Workload Author Seeds Shared Images Into Torrent Separately From the Non-Share Images

- the shared gps image will be broken out into a separate gps.json file that includes the torrent info for it in a section called `torrent_shared` that has the same fields within it as the current `torrent` section.
- the agbot will support the addition of the field called `image_name` in both the torrent.image and torrent_shared.image sections. (This makes it easier for both the workload author and anax to know which tar file is for which image.)
- the agbot will support include file processing for policy files so that gps.json can be included into the location and gpstest policy files.
- the policy file signing scripts will be modified to support seeding the json file, and to sign/seed  the policy files even if they have an include statement in them.
- the result of these changes is that when anax receives both the location and gpstest policy files they will reference the same tar file name for the gps image, so anax will know immediately that it already has it.

**Comment by dabooz** [comment]
Option 1 is cheaper to implement and deliver, and provides a better user experience, so we should go with that.

**Comment by dabooz** [comment]
After looking in depth at option 1 and 2 (and discovering how ugly they are), and then a discussion with @bmpotter we came to the conclusion that this problem with shared container wont exist after we do the proper split of microservices and workloads. All the use cases we have for shared containers are for cases where the shared container is actually a microservice packaged with the workload, and further the reason we need it to be shared today is precisely because we dont have a proper MS-workload split. After we implement a proper split (a) all microservices will be shareable as singletons and (b) there is no danger of shared containers being specified in workload deployment patterns because NO microservice containers are ever specified in workload deployment patterns.

Closing this issue.
